### ISSUE TITLE
[non-nullable-type-assertion-style] triggers on `as const` declaration

### ISSUE THREAD

**Issue by leaumar** [typescript-eslint/typescript-eslint#2879]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/non-nullable-type-assertion-style": ["error"]
  }
}
```

```TS
const x = [] as const;
```

```JSON
{
    "compilerOptions": {
        "inlineSourceMap": false,
        "moduleResolution": "node",
        "noUncheckedIndexedAccess": true,
        "strict": true,
        "target": "es2020"
    },
    "include": ["src/**/*.ts"]
}
```

**Expected Result**

`<anything> as const` declarations shouldn't trigger a "don't do a type cast where `!` is sufficient" warning since `as const` doesn't remove any nullability, just makes everything deep-readonly.

**Actual Result**

The rule thinks `as const` is a typecast removing null from a type:

>Use a ! assertion to more succintly remove null and undefined from the type  @typescript-eslint/non-nullable-type-assertion-style

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.10.0` |
| `@typescript-eslint/parser`        | `4.10.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `14.15.1` |
